### PR TITLE
Added Test for Function Module. New STE operation. Added Documentation

### PR DESCRIPTION
**Pull request by Giuseppe5**
Added tests for operations in Function module.
Re-wrote the STE operator using C++ APIs and JIT for Pytorch >1.3.0, or autograd.Function class in Python without JIT for previous versions.
Wrote documentation for each function tested.

### CODE REVIEW

File: brevitas/csrc/ops.hpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+at::Tensor SampleInput(const at::Tensor& input);
+at::Tensor SampleInput_backward(const at::Tensor& grad);
+
+using namespace at;
+using torch::Tensor;
+using torch::autograd::AutogradContext;
+using torch::autograd::Variable;
+using torch::autograd::variable_list;
+
+class Round_Ste_Class : public torch::autograd::Function<Round_Ste_Class> {
```
Comment (volcacius) [inline brevitas/csrc/ops.hpp]: CamelCase for class names plz, not Camel_Case

File: brevitas/csrc/ops.hpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
+#include <ATen/TensorUtils.h>
+#include <torch/extension.h>
+
+at::Tensor SampleInput(const at::Tensor& input);
```
Comment (volcacius) [inline brevitas/csrc/ops.hpp]: This looks like it shouldn't be here

File: brevitas/csrc/ops_register.hpp [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+
+#include <Python.h>
+#include <torch/script.h>
+#include "ops.hpp"
```
Comment (volcacius) [inline brevitas/csrc/ops_register.hpp]: Fix newline

File: brevitas/function/autograd_ops.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import torch
+
+
+class scalar_clamp_ste_fn(torch.autograd.Function):
+    """ Python class to implement Straight Trough Estimator operations
+
+    Look at the documentation of `scalar_clamp_ste` for further details.
```
Comment (volcacius) [inline brevitas/function/autograd_ops.py]: Can we find a way to link to the docs of the referenced function through Sphinx?

File: brevitas/function/autograd_ops.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import torch
+
+
+class scalar_clamp_ste_fn(torch.autograd.Function):
+    """ Python class to implement Straight Trough Estimator operations
```
Comment (volcacius) [inline brevitas/function/autograd_ops.py]: I'd say something like "Autograd function that implements scalar_clamp with a straight through estimator".

File: brevitas/function/autograd_ops.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    def backward(ctx, grad_y):
+        """
+        """
+        return grad_y, None, None
+
+
+class tensor_clamp_ste_fn(torch.autograd.Function):
+    """ Python class to implement Straight Trough Estimator operations
+
+    Look at the documentation of `tensor_clamp_ste_fn` for further details.
```
Comment (volcacius) [inline brevitas/function/autograd_ops.py]: I think you meant "tensor_clamp_ste"

File: brevitas/function/ops.py [hunk 7]
```diff
@@ -66,70 +71,44 @@ def tensor_clamp_(x: torch.Tensor, min_val: torch.Tensor, max_val: torch.Tensor)
...
-    return value
-
+def ceil_ste(x: torch.Tensor) -> torch.Tensor:
+    """ Perform ceil operation with Straight Trough Estimation (STE) of the Gradient
 
-@torch.jit.script
-def tensor_clamp_ste(x: torch.Tensor, min_val: torch.Tensor, max_val: torch.Tensor) -> torch.Tensor:
-    y = tensor_clamp(x, min_val, max_val)
-    return _ste(x, y)
+    This operation has no effect on the backpropagation of the gradient. The STE is implemented using the
```
Comment (volcacius) [inline brevitas/function/ops.py]: This operation behaves like an identity on the backward pass.

File: brevitas/__init__.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import imp
+import os
+import torch
+import sys
+from packaging import version
+
+torch_version = version.parse(torch.__version__)
+
+if torch_version >= version.parse("1.3.0"):
```
Comment (volcacius) [inline brevitas/__init__.py]: Add some comments to say what we are doing here please

File: brevitas/csrc/ops.hpp [hunk 9]
```diff
@@ -0,0 +1,169 @@
+#include <ATen/TensorUtils.h>
+#include <torch/extension.h>
+
+using namespace at;
+using torch::Tensor;
+using torch::autograd::AutogradContext;
+using torch::autograd::Variable;
+using torch::autograd::variable_list;
+
+class RoundSteClass : public torch::autograd::Function<RoundSteClass> {
```
Comment (volcacius) [inline brevitas/csrc/ops.hpp]: Can we call it Fn rather than Class? thx

File: brevitas/csrc/ops_register.cpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
+#include "ops_register.hpp"
+// If we are in a Windows environment, we need to define
```
Comment (volcacius) [inline brevitas/csrc/ops_register.cpp]: add white space

File: brevitas/function/autograd_ops.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    Look at the documentation of :func:`~brevitas.function.ops_ste.scalar_clamp_ste` for further details.
+
+    """
+    @staticmethod
+    def forward(ctx, x: torch.Tensor, min_val: float, max_val: float):
+        """
+        """
+        y = torch.clamp(x, min_val, max_val)
+        return y
+    @staticmethod
```
Comment (volcacius) [inline brevitas/function/autograd_ops.py]: whitespace plz

File: setup.py [hunk 12]
```diff
@@ -38,19 +38,144 @@
...
+        with open('.gitignore', 'r') as f:
+            ignores = f.read()
+            for wildcard in filter(None, ignores.split('\n')):
+                for filename in glob.glob(wildcard):
+                    try:
+                        os.remove(filename)
+                    except OSError:
+                        shutil.rmtree(filename, ignore_errors=True)
+
+        # It's an old-style class in Python 2.7...
```
Comment (volcacius) [inline setup.py]: ? We don't support Python 2.7

File: test/common.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
+import hypothesis.strategies as st
+
+# Set Constants
+RTOL = 0
+ATOL = 1e-23
+
+# Define custom type of floating point generator.
+# We are never interested in NaN and Infinity. In some case, such as when generating gradients, we may also want to
+# exclude zero. For scale factor, we want only positive numbers
+float_st = st.floats(allow_nan=False, allow_infinity=False, width=32)
```
Comment (volcacius) [inline test/common.py]: Move 32 bits to the top as a constant

File: test/test_ops.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+    maximum = torch.tensor(minmax[1])
+
+    output = scalar_clamp_ste(value, minimum, maximum)
+    expected_output = torch.clamp(value, minmax[0], minmax[1])
+
+    assert ((output >= minimum).all() and (output <= maximum).all())
+    assert (torch.allclose(expected_output, output, RTOL, ATOL))
+
+
+# Same as before, but with two Tensors of Float
```
Comment (volcacius) [inline test/test_ops.py]: Same as before what?

File: test/test_ops.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    else:
+        expected_output = torch.tensor(0.0)
+
+    expected_output = expected_output
+
+    assert (torch.allclose(expected_output, output, RTOL, ATOL))
+
+
+# Requires two floats as maximum and minimum and a tensor of float
+@given(minmax=two_ordered_numbers(), x=list_float_st)
```
Comment (volcacius) [inline test/test_ops.py]: Are we sure this test the cases where there some clamping happening?

File: test/test_ops.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+        expected_output = (2 ** (bit_width - 1)) - 1
+    else:
+        expected_output = (2 ** bit_width) - 1
+    expected_output = expected_output
+
+    assert (torch.allclose(expected_output, output, RTOL, ATOL))
+
+
+# Test different combinations of Narrow Range (True/False), Signed (True/False), and BitWidth (1...8)
+@given(narrow_range=st.booleans(), signed=st.booleans(), bit_width=st.integers(min_value=0, max_value=8))
```
Comment (volcacius) [inline test/test_ops.py]: min_value should be 1. Add the as constants to the top of the file

File: setup.py [hunk 17]
```diff
@@ -38,19 +38,144 @@
...
+
+class build_py(build_py_orig):
+    def find_package_modules(self, package, package_dir):
+        modules = super().find_package_modules(package, package_dir)
+        if torch_version >= version.parse("1.3.0"):
+            file_to_rename = ['ops_ste_o']
+        else:
+            file_to_rename = ['ops_ste_n']
+
+        pacchetti = [(pkg, mod, file,) for (pkg, mod, file,) in modules
```
Comment (volcacius) [inline setup.py]: pacchetti?

File: setup.py [hunk 18]
```diff
@@ -38,19 +38,144 @@
...
+TEST_REQUIRES = ["pytest", "hypothesis", "mock"]
 
 
 def read(fname):
     return open(os.path.join(os.path.dirname(__file__), fname)).read()
 
 
+if torch_version >= version.parse("1.3.0"):
+    cmdclass_dict = {
+        'build_ext': BuildExtension.with_options(no_python_abi_suffix=True),
```
Comment (volcacius) [inline setup.py]: What is no_python_abi_suffix? Comment please

File: brevitas/function/ops_ste_n.py [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+
+    Parameters
+    ----------
+    x : Tensor
+        Tensor on which to apply the round operation
+
+    Returns
+    -------
+    Tensor
+        Tensor after applying round operation. When backpropagating, the gradient will be unaffected by the
```
Comment (volcacius) [inline brevitas/function/ops_ste_n.py]: When backpropagating through this value, a straight through estimator is applied.

File: setup.py (hunk 18 above)
Comment (volcacius) [reply setup.py]: Add a pointer to the reference setup.py

**Comment by volcacius** [comment]
We also have to update the conda env file

File: test/test_ops.py (hunk 15 above)
Comment (Giuseppe5) [reply test/test_ops.py]: Leaving this as memo. The only constraint set is that min_value < max_value, while there are no constraints on the value in x. Considering that the number are randomly generated, there is high chance that the clamp operation is really executed. A similar thing can be said and extended for the tensor_clamp case.

**Pull request merged**